File: app/src/main/java/io/blackbox_vision/helpers/logic/presenter/DetailsPresenter.java
package io.blackbox_vision.helpers.logic.presenter;

import android.os.Bundle;
import android.support.annotation.NonNull;

import io.blackbox_vision.helpers.logic.view.DetailsView;
import io.blackbox_vision.helpers.logic.interactor.DetailsInteractor;
import io.blackbox_vision.mvphelpers.logic.presenter.BasePresenter;


public final class DetailsPresenter extends BasePresenter<DetailsView> {
    private static DetailsPresenter detailsPresenter = null;
    private DetailsInteractor detailsInteractor;

    private DetailsPresenter() { }

    @Override
    public void onViewAttached(@NonNull DetailsView view) {
        //Init your interactors here
        detailsInteractor = DetailsInteractor.newInstance();

        //Restore app state
    }

    @Override
    public void onViewDetached() {
        //Dereference interactors
        detailsInteractor = null;

        //Save app state
    }

    public void findRequiredInformation(@NonNull String id) {
        if (isViewAttached()) {
            detailsInteractor.retrieveDetailsFromService(id, this::onSuccess, this::onError);
        }
    }

    private void onSuccess(@NonNull Bundle information) {
        if (isViewAttached()) {
            getView().onInfoReceived(information);
        }
    }

    private void onError(@NonNull String errorMessage) {
        if (isViewAttached()) {
            getView().onInfoError(errorMessage);
        }
    }

    public static DetailsPresenter newInstance() {
        if (null == detailsPresenter) {
            detailsPresenter = new DetailsPresenter();
        }

        return detailsPresenter;
    }
}

File: app/src/main/java/io/blackbox_vision/helpers/logic/presenter/DetailsPresenter.kt
package io.blackbox_vision.helpers.logic.presenter

import android.os.Bundle

import io.blackbox_vision.helpers.logic.view.DetailsView
import io.blackbox_vision.helpers.logic.interactor.DetailsInteractor
import io.blackbox_vision.mvphelpers.logic.listener.OnErrorListener
import io.blackbox_vision.mvphelpers.logic.listener.OnSuccessListener
import io.blackbox_vision.mvphelpers.logic.presenter.BasePresenter

class DetailsPresenter : BasePresenter<DetailsView>(), OnSuccessListener<Bundle>, OnErrorListener<String> {
    private val interactor: DetailsInteractor

    init {
        interactor = DetailsInteractor()
    }

    fun findRequiredInformation(id: String) {
        if (isViewAttached) {
            interactor.retrieveDetailsFromService(id, this, this)
        }
    }

    override fun onSuccess(data: Bundle) {
        if (isViewAttached) {
            view!!.onInfoReceived(data)
        }
    }

    override fun onError(error: String) {
        if (isViewAttached) {
            view!!.onInfoError(error)
        }
    }
}
